### PR TITLE
apps/examples: Add test case for stack overflow protection.

### PR DESCRIPTION
**Pull request by sangam-swami**
Add test case to verify stack overflow protection feature.

Signed-off-by: Sangamanatha <sangam.swami@samsung.com>

### CODE REVIEW

File: apps/examples/stack_protection/stack_protection_main.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	return 0;
+}
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int stack_prot_tc_main(int argc, char *argv[])
+#endif
+{
+
```
Comment (jeongchanKim) [inline apps/examples/stack_protection/stack_protection_main.c:107]: Please remove unnecessary empty line.

File: apps/examples/stack_protection/stack_protection_main.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	pthread_t tid;
+
+	printf("Stack Overflow Protection testcase!!\n");
+	/* Create Thread for testing */
+	tid = create_stack_tc_thread();
+	if (tid < 0) {
+		return EXIT_FAILURE;
+	}
+
+	pthread_join(tid, NULL);
```
Comment (jeongchanKim) [inline apps/examples/stack_protection/stack_protection_main.c:117]: Is this really needed? I think, you already wait(pthread_join) in create_stack_tc_thread(), and there is no reason that this function should wait the termination of pthread.

File: apps/examples/stack_protection/stack_protection_main.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	ret = pthread_create(&tid, &attr, stack_tc_thread, NULL);
+	if (ret != 0) {
+		fprintf(stderr, "failed to create pthread(%d)\n", ret);
+		return -ret;
+	}
+
+	pthread_setname_np(tid, "stack tc thread");
+	pthread_join(tid, NULL);
+	return 0;
```
Comment (jeongchanKim) [inline apps/examples/stack_protection/stack_protection_main.c:98]: I think you'd better return ret instead of 0.

File: apps/examples/stack_protection/stack_protection_main.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+ ****************************************************************************/
+
+static pthread_addr_t stack_tc_thread()
+{
+	char ch;
+	int i = 0;
+	volatile char array[STACK_THREAD_SIZE];
+
+	while (1) {
+		printf("\nPress N - Test Stack access within range\n");
```
Comment (jeongchanKim) [inline apps/examples/stack_protection/stack_protection_main.c:46]: There is no way for exit the test without overflow assertion.
How about adding new option for normal termination of test?

File: apps/examples/stack_protection/Makefile [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+############################################################################
+############################################################################
+# apps/examples/stack_overflow/Makefile
```
Comment (amandeepc) [inline apps/examples/stack_protection/Makefile:20]: Please correct file path.
stack_overflow -> stack_protection

Comment (sangam-swami) [reply apps/examples/stack_protection/Makefile]: yes changed

File: apps/examples/stack_protection/stack_protection_main.c (hunk 4 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c:51]: yes added test case exit option

File: apps/examples/stack_protection/stack_protection_main.c (hunk 3 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: yes handled

File: apps/examples/stack_protection/stack_protection_main.c (hunk 1 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: changed

File: apps/examples/stack_protection/Kconfig [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_STACK_PROTECTION
+	bool "Stack Protection TestCase Example"
```
Comment (sunghan-chang) [inline apps/examples/stack_protection/Kconfig:7]: It does not have the same format with another testcase which is in tc folder.
Let's say `Stack Protection Test example`

File: apps/examples/stack_protection/stack_protection_main.c [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+ * Pre-processor Definitions
+ ****************************************************************************/
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int stack_prot_tc_main(int argc, char *argv[])
```
Comment (sunghan-chang) [inline apps/examples/stack_protection/stack_protection_main.c:40]: stack_prot_main

File: apps/examples/stack_protection/stack_protection_main.c [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+#endif
+{
+	char ch;
+	int i = 0;
+	volatile char array_normal[10];
+	volatile char array_overflow[CONFIG_TASH_CMDTASK_STACKSIZE];
```
Comment (sunghan-chang) [inline apps/examples/stack_protection/stack_protection_main.c:46]: 1. You set 512 for stacksize in makefile so that `CONFIG_TASH_CMDTASK_STACKSIZE` is not used. And this should be working with TASH disabled.
Please use another value.
2. Let's use new definitions for those values. 10 -> xxx, config_tash_xx -> xxxxx

File: apps/examples/stack_protection/stack_protection_main.c [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+			printf(" Test stack access in overflow - Failed !!\n");
+			sleep(1);
+		} else if (ch == 'E' || ch == 'e') {
+			printf(" Exit the Test ..\n");
+			sleep(1);
+			return EXIT_SUCCESS;
+		}
+	}
+
+	sleep(1);
```
Comment (sunghan-chang) [inline apps/examples/stack_protection/stack_protection_main.c:74]: Let's remove unnecessary code.

File: apps/examples/stack_protection/stack_protection_main.c [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <sys/types.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
```
Comment (sunghan-chang) [inline apps/examples/stack_protection/stack_protection_main.c:28]: unnecessary

File: apps/examples/stack_protection/Kconfig [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+	depends on MPU_STACK_OVERFLOW_PROTECTION
+	---help---
+		Enable the Stack Overflow Protection TestCase Example
+
+config USER_ENTRYPOINT
+	string
+	default "stack_protection_main" if ENTRY_STACK_PROTECTION
```
Comment (amogh-samsung) [inline apps/examples/stack_protection/Kconfig:14]: Default entry point name not in line with C file function name `stack_prot_tc_main`

File: apps/examples/stack_protection/Kconfig (hunk 6 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/Kconfig]: added to example test case for support in tash and non tash case, let me know if that is fine.

File: apps/examples/stack_protection/stack_protection_main.c (hunk 7 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: changed

File: apps/examples/stack_protection/stack_protection_main.c (hunk 9 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: ok

File: apps/examples/stack_protection/stack_protection_main.c (hunk 10 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: removed

File: apps/examples/stack_protection/Kconfig (hunk 11 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/Kconfig]: changed

File: apps/examples/stack_protection/stack_protection_main.c (hunk 8 above)
Comment (sangam-swami) [reply apps/examples/stack_protection/stack_protection_main.c]: 1. Changed the stack size to 1024.
2. yes changed to macros
